File: stronghold/001_DNA/ocaml/dna.ml
(* Count nucleotides in a DNA sequence *)
let count_nucs = fun dna -> 
    let counts = Hashtbl.create 4 in
    (* Initializes the counts for ACGT *)
    List.iter (fun c -> Hashtbl.add counts c 0) ['A'; 'C'; 'G'; 'T'];

    String.iter (fun c -> 
        if Hashtbl.mem counts c then 
            Hashtbl.replace counts c ((Hashtbl.find counts c) + 1)
        ) dna; 

    (
        Hashtbl.find counts 'A', 
        Hashtbl.find counts 'C', 
        Hashtbl.find counts 'G', 
        Hashtbl.find counts 'T'
    ) ;;

let () = 
    let seq = read_line () in
    let (a, c, g, t) = count_nucs seq in 
    Printf.printf "%d %d %d %d\n" a c g t ;;

File: stronghold/001_DNA/lua/dna.lua
---increments count for base base in table tbl
---@param tbl table
---@param base string
---@return nil
local function increase_base_count (tbl, base)
    if base == "A" then tbl.a = tbl.a + 1
    elseif base == "C" then tbl.c = tbl.c + 1
    elseif base == "G" then tbl.g = tbl.g + 1
    elseif base == "T" then tbl.t = tbl.t + 1
    end
    return nil
end

---produces a table with counts of standard nucleotides (ACGT) in a sequence
---@param sequence string
---@return table
function CountNucleotides(sequence)
    local out = {a = 0, c = 0, g = 0, t = 0}
    for i = 1, #sequence do
        local base = sequence:sub(i, i)
        increase_base_count(out, base)
    end
    return out
end

-- file execution
local dna = io.read("*a"):gsub("%s+", "")
local bc = CountNucleotides(dna) -- short for base counts
print(string.format("%d %d %d %d", bc.a, bc.c, bc.g, bc.t))
